Cython/Build/Tests/TestRecythonize.py

```diff
@@ -0,0 +1,204 @@
+import shutil
+import os
+import tempfile
+
+import Cython.Build.Dependencies
+import Cython.Utils
+from Cython.TestUtils import CythonTest
+
+
+
+SAME = "The result of cytonization is the same"
+INCOR = "Incorrect cythonization"
+LINE_1 = '  /* "{name}{ext}":1\n'
+VARS_LINE = '  /*--- Wrapped vars code ---*/\n'
+
+
+class TestRecythonize(CythonTest):
+
+    def setUp(self):
+        CythonTest.setUp(self)
+        self.temp_dir = (
+            tempfile.mkdtemp(
+                prefix='recythonize-test',
+                dir='TEST_TMP' if os.path.isdir('TEST_TMP') else None
+            )
+        )
+        self.src_dir = tempfile.mkdtemp(prefix='src', dir=self.temp_dir)
+        self.dep_tree = Cython.Build.Dependencies.create_dependency_tree()
+        self.language_level = 3
+
+    def tearDown(self):
+        CythonTest.tearDown(self)
+        shutil.rmtree(self.temp_dir)
+
+    def fresh_cythonize(self, *args, **kwargs):
+        Cython.Utils.clear_function_caches()
+        Cython.Build.Dependencies._dep_tree = None  # discard method caches
+        kwargs.update(language_level=self.language_level)
+        Cython.Build.Dependencies.cythonize(*args, **kwargs)
+
+    def refresh_dep_tree(self):
+        Cython.Utils.clear_function_caches()
+        Cython.Utils.clear_method_caches(self.dep_tree)
+        self.dep_tree._transitive_cache.clear()
+
+    def fresh_all_dependencies(self, *args, **kwargs):
+        self.refresh_dep_tree()
+        return self.dep_tree.all_dependencies(*args, **kwargs)
+
+    def relative_lines(self, lines, line, start, end):
+        try:
+            ind = lines.index(line)
+            return lines[ind+start: ind+end]
+        except ValueError:
+            # XXX: It is assumed that VARS_LINE is always present.
+            ind = lines.index(VARS_LINE)
+            raise ValueError(
+                "{0!r} was not found, presumably in {1}".format(
+                    line, lines[ind-10: ind-1]))
+        except Exception as e:
+            raise e
+
+    def relative_lines_from_file(self, path, line, start, end):
+        with open(path) as f:
+            lines = f.readlines()
+
+        return self.relative_lines(lines, line, start, end)
+
+    def recythonize_on_pxd_change(self, ext, creating_pxd):
```

Comment:
> Rename it to "create_module_pxd" and make it create a module pxd file if True.

In general, this name of the argument is not very specific about the behavior, since in both cases pxd is created. I will try to reflect the behavior more clearly in resolution.

Reply:
resolved [here](https://github.com/cython/cython/pull/4241/commits/059a5f883d787a5b2664dcd79e43fe12a7730d29)